yb-voyager/cmd/endMigrationCommand.go

```diff
@@ -0,0 +1,350 @@
+		utils.ErrExit("end migration: initializing target db: %v", err)
+	}
+	defer tdb.Finalize()
+	tdb.ClearMigrationState(migrationUUID, exportDir)
+}
+
+func cleanupFallForwardDB(msr *metadb.MigrationStatusRecord) {
+	if msr.FallForwardEnabled {
+		utils.PrintAndLog("cleaning up fall-forward db...")
+		ffdb := tgtdb.NewTargetDB(msr.FallForwardDBConf)
+		err := ffdb.Init()
+		if err != nil {
+			utils.ErrExit("end migration: initializing fallforward db: %v", err)
+		}
+		defer ffdb.Finalize()
+		ffdb.ClearMigrationState(migrationUUID, exportDir)
+	}
+}
+
+func cleanupFallBackDB(msr *metadb.MigrationStatusRecord) {
+	if msr.FallbackEnabled {
+		utils.PrintAndLog("cleaning up fallback db...")
+		fbdb := tgtdb.NewTargetDB(msr.SourceDBAsTargetConf)
+		err := fbdb.Init()
+		if err != nil {
+			utils.ErrExit("end migration: initializing fallback db: %v", err)
+		}
+		defer fbdb.Finalize()
+		fbdb.ClearMigrationState(migrationUUID, exportDir)
+	}
+}
+
+func cleanupExportDir() {
+	utils.PrintAndLog("cleaning up export dir...")
+	subdirs := []string{"schema", "data", "logs", "reports", "temp", "metainfo"}
+	for _, subdir := range subdirs {
+		err := os.RemoveAll(filepath.Join(exportDir, subdir))
+		if err != nil {
+			utils.ErrExit("end migration: removing %s directory: %v", subdir, err)
+		}
+	}
+}
+
+func backupSchemaFilesFn(msr *metadb.MigrationStatusRecord) {
+	if !bool(backupSchemaFiles) {
+		return
+	}
+
+	log.Infof("backing up schema files to %s", backupDir)
+	err := os.MkdirAll(filepath.Join(backupDir, "schema-backup"), 0755)
+	if err != nil {
+		utils.ErrExit("end migration: creating schema backup directory: %v", err)
+	}
+
+	cmd := exec.Command("mv", filepath.Join(exportDir, "schema"), filepath.Join(backupDir, "schema-backup"))
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		utils.ErrExit("end migration: moving schema files: %s: %v", string(output), err)
+	}
+}
+
+func backupDataFilesFn() {
+	if !bool(backupDataFiles) {
+		return
+	}
+
+	log.Infof("backing up sql data files to %s", backupDir)
+	err := os.MkdirAll(filepath.Join(backupDir, "data-backup"), 0755)
+	if err != nil {
+		utils.ErrExit("end migration: creating data backup directory: %v", err)
+	}
+
+	files, err := os.ReadDir(filepath.Join(exportDir, "data"))
+	if err != nil {
+		utils.ErrExit("end migration: reading data directory: %v", err)
+	}
+	for _, file := range files {
+		if file.IsDir() || !strings.HasSuffix(file.Name(), ".sql") {
+			continue
+		}
+
+		err = os.Rename(filepath.Join(exportDir, "data", file.Name()), filepath.Join(backupDir, "data-backup", file.Name()))
+		if err != nil {
+			utils.ErrExit("end migration: moving data files: %v", err)
+		}
+	}
+}
+
+func saveMigrationReportsFn() {
+	if !bool(saveMigrationReports) {
+		return
+	}
+	log.Infof("saving up migration reports to %s", backupDir)
+	// TODO: what if there is report.txt generated from analyze-schema step
+	err := os.Rename(filepath.Join(exportDir, "reports"), filepath.Join(backupDir, "reports"))
```

Comment:
What if the `end migration` moved the reports directory and then failed anywhere during the remaining run? If the user re-runs the command, will this statement fail with `reports` directory not exists error?

IMPORTANT NOTE: the entire `end migration` command implementation MUST be idempotent. If the command fails at any point, user should be able to run the command again. The `end migration` command does not fail if previous invocation of `end migration` was successful for the same `exportDir`. Revisit the entire implementation to ensure the idempotency requirement is satisfied.